### PR TITLE
Create more wheels (manylinux and Mac OS).

### LINKED ISSUE #32: Create more wheels
Currently there is only ~~an sdist~~ the sdist and win wheels on PyPI. Having more wheels there would be nice.

### PR DESCRIPTION
**Pull request by icemac**
Additionally add coverage/coveralls support.
The changes are taken from zope.interface.

Fixes #32.

### CODE REVIEW

**Comment by icemac** [comment]
Hm, how do I check that creating the wheels actually works? Do I have to create a new tag for that?

File: .manylinux-install.sh [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+# Compile wheels
+for PYBIN in /opt/python/*/bin; do
+    if [[ "${PYBIN}" == *"cp27"* ]] || \
+       [[ "${PYBIN}" == *"cp34"* ]] || \
+       [[ "${PYBIN}" == *"cp35"* ]] || \
+       [[ "${PYBIN}" == *"cp36"* ]]; then
+        "${PYBIN}/pip" install -e /io/
+        "${PYBIN}/pip" wheel /io/ -w wheelhouse/
+    rm -rf /io/build /io/*.egg-info
```
Comment (mgedmin) [inline .manylinux-install.sh]: Indentation seems wrong here: the rm is inside the `if`, but is indented as if it's outside.

File: setup.cfg [hunk 2]
```diff
@@ -1,7 +1,14 @@
-[nosetests]
-nocapture=1
-cover-package=zodbpickle
-cover-erase=1
-with-doctest=0
-where=src/zodbpickle
+[bdist_wheel]
+universal = 0
 
+[coverage:run]
```
Comment (jamadden) [inline setup.cfg]: Personally I prefer to keep coverage configuration in .coveragerc, but my reasons are pretty trivial. First, `coveralls` passes `".coveragerc"` as the 'config_file' option to `coverage`, as perhaps other tools might. In the long-long ago, `coverage` would only look in that passed file for its config, so it was just generally simpler to keep it there. Second, I find it easier to cargo-cult around a .coveragerc file from project to project---it's just a 'cp' if the file specifies `source = src/`---than it is a 'setup.cfg' if the destination project already has a 'setup.cfg', which is likely.

File: .travis.yml [hunk 3]
```diff
@@ -1,15 +1,76 @@
...
+        - os: linux
+          python: 3.4
+        - os: linux
+          python: 3.5
+        - os: linux
+          python: 3.6
+        - os: linux
+          python: pypy
+        - os: linux
+          python: pypy3
```
Comment (mgedmin) [inline .travis.yml]: TBH I think you can leave the old python: ... list; the matrix: include: bits will be added to it.

Look at [zope.proxy](https://github.com/zopefoundation/zope.proxy/blob/master/.travis.yml) for an example (or zope.component, or zope.container).

**Comment by mgedmin** [comment]
I have _some_ ideas for testing:

- Create a temporary branch and edit .travis.yml to do the build unconditionally, but skip the upload.  This works for Appveyor, where you can designate some files to be archived as artifacts you can then download from Appveyor.  Perhaps Travis has a similar feature?

- Create a temporary branch and edit .travis.yml to do the build unconditionally and tell twine to upload it to the test PyPI server.

- Create a temporary branch based on the last release tag and edit .travis.yml to do the build unconditionally and upload it to the real PyPI server -- this works well for building wheels of older releases.

Be sure to delete the temporary branch from GitHub afterwards.

**Comment by icemac** [comment]
@jamadden wrote: 
> I don't like a `tox` environment that does `coverage erase` and `coverage combine` because it breaks under `detox`.

I do not use `detox` because I got strange errors when trying it on several projects (errors which do not happen when running plain `tox`. 

Why/how does my change break on `detox`? Looking at https://hynek.me/articles/testing-packaging/ it seems be possible to have only one coverage-report section. IMHO it should not matter calling `coverage erase` as the first step when running `coverage-report` as the last step as described by Hynek Schlawack.

> I *would* like the coverage environment to specify a percentage with `--fail-under` so it's doing something useful, not just for human viewing.

:+1: I am adding one.

> The coveralls.io data looks weird and disorganised. Maybe this new "tree" feature they suddenly seem to have is buggy (or I don't understand it) or maybe we need to install and then run the tests (i.e., stop using 'setup.py test')?

Maybe this is because the first run was without a proper coverage config. Let's see what happens after the merge.

> I don't know much about the wheel building parts, but ISTM that if we're going to be doing that, we need an appveyor.yml that does the same---it's Windows users who tend to have the most difficulty building C extensions in my experience.

Windows wheels are already there. :)

> Anyway, it's all better than what we had, so basically LGTM.

Thank you.

File: .travis.yml (hunk 3 above)
Comment (icemac) [reply .travis.yml]: Okay, let's try it.

**Comment by jamadden** [comment]
> I do not use detox because I got strange errors when trying it on several projects (errors which do not happen when running plain tox.

Then those projects are broken 😄 There are some legit reasons why projects can't be tested in parallel (gevent is one---it spawns a number of servers that need to use fixed network ports, which obviously can conflict in parallel), but in my experience most often it's just poorly written tests that assume constant paths to temp files and such.

> Why/how does my change break on detox?

I guess "break" is a strong word. It's just pointless under detox. All the environments listed in `envlist` are run at the same time, so the coverage environment has nothing to report because everything else is still running. At best you get partial data, depending on exactly when 'coverage erase' and 'coverage combine' happen with respect to everything else.

I understand that this project has large swaths of code that are only run under either python 2 or python 3, so having one single environment produce the canonical coverage report is not really tractable. Luckily, Coverage 4.5 was [recently released](https://coverage.readthedocs.io/en/coverage-4.5.1/changes.html#version-4-5-2018-02-03), and its key feature [is designed to address exactly this situation](https://nedbatchelder.com/blog/201802/coverage_45.html), so that both Python 2 and Python 3 can run coverage, and Python 3 can explicitly say "ignore the Python 2 files" and vice versa, allowing each individual environment (theoretically) to have 100% coverage of its own relevant code.

**Comment by jamadden** [comment]
> Windows wheels are already there. :)

Argh, I completely missed that file. Sorry!

**Comment by jamadden** [comment]
[TIL](https://bitbucket.org/ned/coveragepy/issues/563/platform-specific-configuration) that environment variables are expanded in the coverage config, so it should be possible to write a .coveragerc that properly excludes the irrelevant files/lines with a bit of work, even before coverage 4.5.

**Comment by mgedmin** [comment]
> > Windows wheels are already there. :)

> Argh, I completely missed that file. Sorry!

You're the third person to miss it :)

(And I was the one who added it in the first place, and then forgot!)

**Pull request merged**